
## Table of Contents

1. [Problem Statement](#section1)<br>
2. [Importing Packages](#section2)<br>
3. [Loading Data](#section3)<br>
  - 3.1 [Description of the Datasets](#section301)<br>
  - 3.2 [Pandas Profiling before Data Preprocessing](#section302)<br>
4. [Data Preprocessing](#section4)<br>
  - 4.1 [Data Normalization](#section401)<br>
  - 4.2 [Pandas Profiling after Data Preprocessing](#section402)<br>
5. [Exploratory Data Analysis](#section5)<br>
  - 5.1 [**Analysis based on Outlet metrics**](#section501)<br>

<a id=section1></a>
## 1. Problem Statement
 
This dataset contain data for more than 9.5K car sale in Ukraine. Most of them are uses cars so it opens the possibility to analyze features related to car operation.


<a id=section2></a>
## 2. Importing packages

In this step  various packages  needed for profiling and charts is imported/installed.

In [20]:
import ssl
import certifi

#urlopen(request, context=ssl.create_default_context(cafile=certifi.where()))

In [58]:
import numpy as np                     

import pandas as pd
pd.set_option('mode.chained_assignment', None)      # To suppress pandas warnings.
pd.set_option('display.max_colwidth',None)           # To display all the data in each column
pd.options.display.max_columns = 50                 # To display every column of the dataset in head()

import warnings
warnings.filterwarnings('ignore')      
import matplotlib.pyplot as plt
%matplotlib inline

import seaborn as sns
sns.set(style='whitegrid', font_scale=1.3, color_codes=True)      # To apply seaborn styles to the plots.

# Install and update plotly using this command to the latest version (higher than version 4.0.0)

#!pip install plotly --trusted-host pypi.python.org packagename
!pip install --trusted-host https://pypi.org/simple/plotly plotly



ERROR: Could not install packages due to an EnvironmentError: Could not find a suitable TLS CA certificate bundle, invalid path: C:/Users/asivaraman/mycert.crt



In [59]:
import os
os.environ['REQUESTS_CA_BUNDLE']= "C:/Users/asivaraman/Anaconda3//lib/site-packages/cacert.pem"
!pip  install plotly

ERROR: Could not install packages due to an EnvironmentError: Could not find a suitable TLS CA certificate bundle, invalid path: C:/Users/asivaraman/Anaconda3//lib/site-packages/cacert.pem



In [56]:
import certifi
certifi.where()

'C:\\Users\\asivaraman\\Anaconda3\\lib\\site-packages\\certifi\\cacert.pem'

<a id=section3></a>
## 3. Loading the data set

Load the dataset using pandas from car sales file and generate the data frame for further analysis.




In [64]:
df_cars = pd.read_excel('CarSales.xlsx')
df_cars.head()


car        price       body  mileage  engV engType registration  year  \
0    Smart  21091.34235      hatch        0   0.9  Petrol          yes  2016   
1   Nissan  20447.15400  crossover        0   1.2  Petrol          yes  2016   
2  Peugeot  19080.75000  crossover        0   1.2  Petrol          yes  2016   
3  Citroen  24050.63700      hatch        0   1.2  Diesel          yes  2016   
4  Renault  17639.46795  crossover        0   1.2  Petrol          yes  2016   

     model  drive  
0  Forfour   rear  
1  Qashqai  front  
2     2008  front  
3       C4  front  
4   Captur  front

<a id=section301></a>
### 3.1 Description of the Dataset

Bsed on the data set columns, below is the decription for each column


| Column Name | Description |
| ------ | ------ |
| Car | Manufacturer brand.  |
| Price | Seller's price in adverstisement(in USD)|
| body | car body type. |
| mileage | as mentioned in advertisement. |
| engV | Rounded Engine volume. |
| engType |Type of fuel("Other" should be treated as NA) |
| registration |Whether car registered in Ukraine or not |
| year |year od production |
| model |Specific model name |
| drive |drive type|



### Intitail screening of the  dataset

Let us try to understand some intital observation on this data based on panda basic panda functions.





In [4]:
df_cars.columns
#df_cars[[df_cars["engV"].astype(float) <1 ].count()
#df_cars[df_cars["engV"].astype(float) <1  | (df_cars["engV"].astype(float) >10 )].count()

Index(['car', 'price', 'body', 'mileage', 'engV', 'engType', 'registration',
       'year', 'model', 'drive'],
      dtype='object')

In [5]:
 #df_cars.columns=map(str.lower, df_cars.columns)


In [6]:
#type of each columns
df_cars.dtypes

car             object 
price           float64
body            object 
mileage         int64  
engV            float64
engType         object 
registration    object 
year            int64  
model           object 
drive           object 
dtype: object

####  Change  car column to brand which is moer relalistic . 


In [7]:
df_cars.rename(columns={'car':"brand"},inplace=True)

In [8]:
df_cars

brand        price       body  mileage  engV engType  \
0     Smart          21091.34235  hatch      0        0.9   Petrol   
1     Nissan         20447.15400  crossover  0        1.2   Petrol   
2     Peugeot        19080.75000  crossover  0        1.2   Petrol   
3     Citroen        24050.63700  hatch      0        1.2   Diesel   
4     Renault        17639.46795  crossover  0        1.2   Petrol   
...       ...                ...        ... ..        ...      ...   
9571  Toyota         36000.00000  crossover  940     NaN    Other    
9572  Mercedes-Benz  68700.00000  crossover  980      3.0   Diesel   
9573  VAZ            750.00000    sedan      999      1.3   Petrol   
9574  Fiat           4000.00000   van        999      2.5   Diesel   
9575  GAZ            0.00000      crossover  1       NaN    Petrol   

     registration  year             model  drive  
0     yes          2016  Forfour           rear   
1     yes          2016  Qashqai           front  
2     yes          2016  2008              front  
3     yes          2016  C4                front  
4     yes          2016  Captur            front  
...   ...           ...     ...              ...  
9571  yes          2009  Land Cruiser 200  NaN    
9572  yes          2013  GL 350            full   
9573  yes          1984  2106              rear   
9574  yes          1998  Ducato ����.      front  
9575  yes          1963  69                full   

[9576 rows x 10 columns]

In [9]:
#find is there any duplicate rows
print("total rows: ",df_cars.shape[0] )
print("dupluate rows : " , df_cars[df_cars.duplicated()]["brand"].count())

total rows:  9576
dupluate rows :  113


#### Out of 9576 columns 113 are duplictes. We have to drop the duplicates now


In [10]:
# remove only the duplicates by keeping first occurence
df_cars.drop_duplicates(keep='first',inplace=True)
df_cars.shape


(9463, 10)

In [11]:
df_cars.describe()
#df_cars

price      mileage         engV         year
count  9463.000000    9463.000000  9029.000000  9463.000000
mean   15303.858896   139.924654   2.641444     2006.525837
std    23057.384098   98.466036    5.963016     7.062584   
min    0.000000       0.000000     0.100000     1953.000000
25%    4999.000000    70.000000    1.600000     2004.000000
50%    9200.000000    129.000000   2.000000     2008.000000
75%    16500.000000   195.000000   2.500000     2012.000000
max    547800.000000  999.000000   99.990000    2016.000000

In [12]:

df_cars.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9463 entries, 0 to 9575
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   brand         9463 non-null   object 
 1   price         9463 non-null   float64
 2   body          9463 non-null   object 
 3   mileage       9463 non-null   int64  
 4   engV          9029 non-null   float64
 5   engType       9463 non-null   object 
 6   registration  9463 non-null   object 
 7   year          9463 non-null   int64  
 8   model         9463 non-null   object 
 9   drive         8953 non-null   object 
dtypes: float64(2), int64(2), object(6)
memory usage: 813.2+ KB


<a id=section302></a>
### 3.2 Pandas Profiling before Data Preprocessing
 we will perform Pandas Profiling before preprocessing our dataset, so we will name the output file as profiling_before_preprocessing_carsales.html.
 

In [63]:

import ssl
orig_sslsocket_init = ssl.SSLSocket.__init__
ssl.SSLSocket.__init__ = lambda *args, cert_reqs=ssl.CERT_NONE, **kwargs: orig_sslsocket_init(*args, cert_reqs=ssl.CERT_NONE, **kwargs)
#!pip install pandas_profiling==1.5.6
!pip --version
!pip install --upgrade pip
#easy_install pip==1.2.1  
#!easy_install pip==1.5.6

pip 20.0.2 from C:\Users\asivaraman\Anaconda3\lib\site-packages\pip (python 3.7)



ERROR: Could not install packages due to an EnvironmentError: Could not find a suitable TLS CA certificate bundle, invalid path: C:/Users/asivaraman/Anaconda3//lib/site-packages/cacert.pem



In [5]:
import pandas_profiling 

ModuleNotFoundError: No module named 'pandas_profiling'

In [ ]:
profile = df_cars.profile_report(title='Pandas Profiling before Data Preprocessing')
profile.to_file(output_file="profiling_before_preprocessing_carsales1.html")

###### Observations from Pandas Profiling before Data Processing
Dataset info

- Number of variables	11
- Number of observations	9463
- Missing cells	944 (0.9%)
- Duplicate rows	0 (0.0%)
- Total size in memory	813.4 KiB
- Average record size in memory	88.0 B
- Variables types

- Numeric	5
- Categorical	5
- Boolean	1



- drive has 510 (5.4%) missing values	Missing : we have to work on it 
-  engv has 434 (4.6%) missing values	also outliers are ther 20,99 etc is not an expected engine capacity.
- mileage has 310 (3.3%) zeros	Zeros : Let us reverify on this .
- model has a high cardinality: 863 distinct values	Warning:   A category with these many distinct values we can delete it
- price has 240 (2.5%) zeros : Zero procing is not acceptible . we have to work on this.


<a id=section4></a>
### 4. Data Preprocessing

<a id=section401></a>
### 4.1 Data Normalization

####  normalize drive column . 

In [ ]:
df_cars.isnull().sum()
drive_mode_val = df_cars["drive"].mode()
print(drive_mode_val,)
# drive got 510 nan values. let us replace it with mode of  drive

#df_cars.describe(include='all')
df_cars["drive"].replace(np.NAN,drive_mode_val,inplace=True)
#df_cars["drive"].mode()


In [ ]:
# check again there is no NaN for drive
df_cars.isnull().sum()

In [ ]:
df_cars["drive"]

####  normalize engv column

Some of the engv values like 20,22...99 and .01,.6 etc is something not real . So let us apply some logic to update these values first to NaN and see how it behaves

In [ ]:
def clean_engv(engv):
    if(engv>=1 or engv<=8 ):
        return  engv
    else:
        return np.nan
df_cars["engv_clean"] =df_cars["engv"].apply(clean_engv)
df_cars.head()
    #print(df_cars["engv_clean"] .min())

In [ ]:
#after applying the clean engv for outliers let us see how many Nan now
df_cars["engv_clean"].isnull().sum()

In [ ]:
#instead of removing this 434  values let us replace them with median of engv_clean
df_cars.isnull().sum()
engv_clean_median = df_cars["engv_clean"].median()
print(engv_clean_median)

df_cars["engv_clean"].replace(np.NAN,engv_clean_median,inplace=True)




In [ ]:
#vrify there is no more nan values for eng_clean
df_cars["engv_clean"].isnull().sum()

In [ ]:
# we can drop engv as we have a cleaned column engv_clean now
#df_cars.drop(['engv'],axis=1,inplace=True)
#df_cars

# Observation on mileage colum
#it ranges fro m 0-999 . 999 0r 0 is not an oulier , because the mileage can  be 0 for unused cars and can be 999 for used cars


### delele model column due to high cardinality for categorical data


In [ ]:
df_cars.drop(['model'],axis=1,inplace=True)
df_cars

In [ ]:
df_cars

In [ ]:
def get_median_mileage(x):
    brand = x.name
    print(x.name,x.count(),x.median()) 
    if x.count() > 0:
        return x.median() # Return median for a brand/model if the median exists. 
engv_median = df_cars.groupby(['brandname'])['engv'].apply(get_median_mileage).reset_index()
#print(mileage_median)
engv_median

#df_cars.groupby(['brandname'])['engv'].median()

In [ ]:
df_cars["engv"].isnull().sum()

In [75]:
df_cars['drive'][df_cars['car'] == 'Renault'].value_counts()

front    436
full      16
rear       2
Name: drive, dtype: int64